File: frontend/src/stores/colony.ts
import { defineStore } from 'pinia'
import { computed, ref } from 'vue'

import { api, ApiRequestError } from '@/api/client'
import type {
  ColonyStateData,
  HysteresisPolicy,
  LogEntry,
  OfflineReport,
  PowerState,
  PopulationState,
  Resources,
  SuspicionState,
} from '@/types/game'
import { DISPLAY_BALANCE, clamp } from '@/utils/balance'
import { clockText } from '@/utils/format'
import { useFacilitiesStore } from '@/stores/facilities'

const EMPTY_RESOURCES: Resources = {
  catnip: 0,
  scrap: 0,
  chips: 0,
  alloys: 0,
  battery: 0,
  lube: 0,
}

/**
 * 基地状态仓（代码结构稿 §3.2.1）。
 *
 * 双值设计：
 * * `serverResources` —— 后端权威值（每 15 秒快照或每次操作后刷新）；
 * * `resources`       —— 100ms 插值的**显示值**，只喂给渲染层；
 * 两者共用后端同一套公式参数（见 utils/balance.ts 的显示镜像说明）。
 */
export const useColonyStore = defineStore('colony', () => {
  const facilitiesStore = useFacilitiesStore()

  const slotId = ref(1)
  const planetId = ref(0)
  const loaded = ref(false)
  const busy = ref(false)
  const lastError = ref<string | null>(null)

  const serverResources = ref<Resources>({ ...EMPTY_RESOURCES })
  const resources = ref<Resources>({ ...EMPTY_RESOURCES })
  const caps = ref<Record<string, number>>({})
  const power = ref<PowerState>({
    gen_kw: 0,
    load_kw: 0,
    net_kw: 0,
    battery_kwh: 0,
    battery_kwh_max: 200,
    blackout: false,
  })
  const population = ref<PopulationState>({
    total: 0,
    max_cap: 0,
    unassigned: 0,
    birth_progress: 0,
  })
  const workstations = ref<Record<string, number>>({})
  const workstationLimits = ref<Record<string, number>>({})
  const facilities = ref<Record<string, number>>({})
  const suspicion = ref<SuspicionState>({ current: 0, max: 100 })
  const policy = ref<HysteresisPolicy | null>(null)
  const offlineReport = ref<OfflineReport | null>(null)
  const lastTickTime = ref(0)
  const lastSyncAt = ref(0)
  const logs = ref<LogEntry[]>([])
  const narrative = ref<string | null>(null)
  const manualClicksLeft = ref<number | null>(null)
  const localBirthProgress = ref(0)
  const localSuspicion = ref(0)
  let logSeq = 0

  function log(text: string, kind: LogEntry['kind'] = 'info') {
    logSeq += 1
    logs.value = [{ id: logSeq, at: clockText(), text, kind }, ...logs.value].slice(0, 80)
  }

  const catnipRate = computed(() => {
    const farmers = workstations.value.farmer ?? 0
    return (
      farmers * facilitiesStore.jobRate('farmer') -
      population.value.total * DISPLAY_BALANCE.catnipConsumePerCatPerSec
    )
  })

  const scrapRate = computed(
    () => (workstations.value.scavenger ?? 0) * facilitiesStore.jobRate('scavenger'),
  )

  const isStarving = computed(
    () => resources.value.catnip <= 0 && catnipRate.value < 0 && population.value.total > 0,
  )

  const coldStartPhase = computed(() => {
    if ((facilities.value.housing_box ?? 0) === 0) return 'GATHER' as const
    if ((facilities.value.farm_plot ?? 0) === 0) return 'FIRST_FARM' as const
    if ((workstations.value.farmer ?? 0) === 0) return 'ASSIGN_FARMER' as const
    if ((facilities.value.scavenge_station ?? 0) === 0) return 'AUTOMATE_SCRAP' as const
    // 有闲置猫口就先上岗（手上有猫比扩容更急）
    if ((workstations.value.scavenger ?? 0) === 0) return 'ASSIGN_SCAVENGER' as const
    // 农夫与拾荒猫都上工 ⇒ 冷启动心流闭环完成，后续交给长线经营
    return 'DONE' as const
  })

  function applyState(data: ColonyStateData) {
    slotId.value = data.slot_id
    planetId.value = data.planet_id
    serverResources.value = {
      catnip: data.resources.catnip,
      scrap: data.resources.scrap,
      chips: data.resources.chips,
      alloys: data.resources.alloys,
      battery: data.resources.battery,
      lube: data.resources.lube,
    }
    resources.value = { ...serverResources.value }
    caps.value = data.resources.caps ?? {}
    power.value = data.power
    population.value = data.population
    workstations.value = data.workstations
    workstationLimits.value = data.workstation_limits
    facilities.value = data.facilities
    suspicion.value = data.suspicion
    offlineReport.value = data.offline_report
    lastTickTime.value = data.last_tick_time
    lastSyncAt.value = Date.now()
    localBirthProgress.value = data.population.birth_progress
    localSuspicion.value = data.suspicion.current
    loaded.value = true
  }

  /** 100ms 高频时钟：只做显示插值，绝不回写权威值 */
  function tick(dtSeconds: number) {
    if (!loaded.value) return
    const next = { ...resources.value }
    next.catnip = clamp(next.catnip + catnipRate.value * dtSeconds, 0, caps.value.catnip ?? 200)
    next.scrap = clamp(next.scrap + scrapRate.value * dtSeconds, 0, caps.value.scrap ?? 200)
    resources.value = next

    const { total, max_cap } = population.value
    if (total > 0 && max_cap > 0 && total < max_cap) {
      const r = DISPLAY_BALANCE.breedingRate
      localBirthProgress.value += r * total * (1 - total / max_cap) * dtSeconds
    }

    const activeFacilities = Object.values(facilities.value).filter((level) => level > 0).length
    const rate =
      DISPLAY_BALANCE.suspicionBaseNoise +
      DISPLAY_BALANCE.suspicionPerActiveFacility * activeFacilities -
      DISPLAY_BALANCE.suspicionIdleDecay
    localSuspicion.value = clamp(
      localSuspicion.value + rate * dtSeconds,
      0,
      suspicion.value.max,
    )
  }

  async function refresh(options: { silent?: boolean; report?: boolean } = {}) {
    try {
      const envelope = await api.getState(slotId.value)
      applyState(envelope.data)
      if (!options.silent) {
        const report = envelope.data.offline_report
        if (report && report.elapsed_seconds > 0) {
          log(
            `读档完成：离线 ${Math.round(report.elapsed_seconds)} 秒，猫薄荷 ${report.gained_catnip.toFixed(1)}、废铁 ${report.gained_scrap.toFixed(1)}、新猫 ${report.gained_cats}`,
          )
          for (const note of report.notes) log(note, 'warn')
        }
      }
      lastError.value = null
    } catch (error) {
      handleError(error, '读档失败')
    }
  }

  function handleError(error: unknown, prefix: string) {
    if (error instanceof ApiRequestError) {
      const text = error.detail ? `${prefix}：${error.message}（${error.detail}）` : `${prefix}：${error.message}`
      lastError.value = text
      log(text, 'crit')
    } else {
      lastError.value = String(error)
      log(`${prefix}：${String(error)}`, 'crit')
    }
    window.setTimeout(() => {
      lastError.value = null
    }, 5000)
  }

  async function scavenge() {
    busy.value = true
    try {
      const envelope = await api.scavenge(slotId.value)
      const data = envelope.data
      resources.value = { ...resources.value, scrap: data.scrap }
      serverResources.value = { ...serverResources.value, scrap: data.scrap }
      manualClicksLeft.value = data.clicks_left
      log(`翻检地表建筑废墟：+1 机械废铁（累计 ${data.manual_scavenge_clicks} 次）`)
      if (data.hint) log(data.hint, 'warn')
      await refresh({ silent: true })
    } catch (error) {
      handleError(error, '翻找失败')
    } finally {
      busy.value = false
    }
  }

  async function build(facilityId: string, count = 1) {
    busy.value = true
    try {
      const envelope = await api.build({ slot: slotId.value, facility_id: facilityId, count })
      const data = envelope.data
      log(`${data.facility_id} 建造完成 → Lv.${data.level}（花费 ${JSON.stringify(data.cost_paid)}）`)
      if (data.narrative) {
        narrative.value = data.narrative
        log('一只折耳流浪猫入驻了第一个纸箱窝！', 'info')
      }
      await refresh({ silent: true })
    } catch (error) {
      handleError(error, '建造失败')
    } finally {
      busy.value = false
    }
  }

  async function dispatch(role: string, delta: number, nextPolicy?: HysteresisPolicy) {
    busy.value = true
    try {
      const envelope = await api.dispatch({
        slot: slotId.value,
        role,
        delta,
        ...(nextPolicy ? { policy: nextPolicy } : {}),
      })
      const data = envelope.data
      if (nextPolicy) policy.value = nextPolicy
      log(`工位调度：${role} ${delta > 0 ? '+' : ''}${delta} → ${data.count} 只（空闲 ${data.unassigned}）`)
      await refresh({ silent: true })
    } catch (error) {
      handleError(error, '调度失败')
    } finally {
      busy.value = false
    }
  }

  /** 15 秒静默快照：提交前端预测值（后端重算为准，偏差 > 0.5% 会被记警告） */
  async function snapshot() {
    try {
      const result = await api.snapshot({
        slot: slotId.value,
        planet_id: planetId.value,
        client_time: Date.now() / 1000,
        resources: { ...resources.value },
        population: {
          total: population.value.total,
          birth_progress: Number(localBirthProgress.value.toFixed(4)),
        },
        workstations: { ...workstations.value },
      })
      lastSyncAt.value = Date.now()
      return result.message
    } catch (error) {
      handleError(error, '快照失败')
      return null
    }
  }

  function dismissNarrative() {
    narrative.value = null
  }

  function clearLogs() {
    logs.value = []
  }

  return {
    slotId,
    planetId,
    loaded,
    busy,
    lastError,
    resources,
    serverResources,
    caps,
    power,
    population,
    workstations,
    workstationLimits,
    facilities,
    suspicion,
    policy,
    offlineReport,
    lastTickTime,
    lastSyncAt,
    logs,
    narrative,
    manualClicksLeft,
    localBirthProgress,
    localSuspicion,
    catnipRate,
    scrapRate,
    isStarving,
    coldStartPhase,
    applyState,
    tick,
    refresh,
    scavenge,
    build,
    dispatch,
    snapshot,
    dismissNarrative,
    clearLogs,
    log,
  }
})

File: backend/app/schemas/colony.py
"""基地状态、工位调度与快照对账的请求/响应契约（代码结构稿 §4.3）。"""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field


class SnapshotRequest(BaseModel):
    """`POST /colony/snapshot` 请求体：前端 Pinia 状态快照（含前台计算校验值）。

    * 携带前端预测值只用于**对账告警**；后端一律用 `last_tick_time` 自行重算；
    * 允许携带额外字段（`extra="allow"`），便于前端渐进式扩展而不用改接口。
    """

    model_config = ConfigDict(extra="allow")

    slot: int = Field(default=1, ge=1, le=3, description="存档槽位 1~3")
    planet_id: int | None = Field(default=None, description="星球 ID，缺省取存档的活跃星球")
    client_time: float | None = Field(default=None, description="前端本地时间戳（仅用于诊断）")
    resources: dict[str, float] = Field(default_factory=dict, description="前端预测的资源存量")
    population: dict[str, float] = Field(default_factory=dict, description="前端预测的猫口与繁育进度")
    workstations: dict[str, int] = Field(default_factory=dict, description="前端预测的工位分桶")
    facilities: dict[str, int] = Field(default_factory=dict, description="前端预测的设施等级")


class ResourcesBlock(BaseModel):
    catnip: float = 0.0
    scrap: float = 0.0
    chips: float = 0.0
    alloys: float = 0.0
    battery: float = 0.0
    lube: float = 0.0
    caps: dict[str, float] = Field(default_factory=dict)


class PowerBlock(BaseModel):
    gen_kw: float = 0.0
    load_kw: float = 0.0
    net_kw: float = 0.0
    battery_kwh: float = 0.0
    battery_kwh_max: float = 0.0
    blackout: bool = False


class PopulationBlock(BaseModel):
    total: int = 0
    max_cap: int = 0
    unassigned: int = 0
    birth_progress: float = 0.0


class SuspicionBlock(BaseModel):
    current: float = 0.0
    max: float = 100.0


class SecurityBlock(BaseModel):
    """三级安防预案状态（模块 F）。"""

    decoy_count: int = 0
    cooldown_until: int | None = None
    cooldown_left_seconds: int = 0
    go_dark: bool = False
    policy: dict[str, bool] = Field(default_factory=dict)


class LaunchSiloStage(BaseModel):
    stage: int
    name: str
    cost: dict[str, float] = Field(default_factory=dict)
    done: bool = False
    blocked: bool = False


class LaunchSiloBlock(BaseModel):
    """火箭垂直发射井（模块 K）阶段视图。"""

    level: int = 0
    max_level: int = 4
    stages: list[LaunchSiloStage] = Field(default_factory=list)
    next_stage: LaunchSiloStage | None = None
    can_advance: bool = True
    fortress_down: bool = False
    launched: bool = False


class OfflineReport(BaseModel):
    """《离线休整报表》（A-1 ~ A-6 的可观测字段）。"""

    elapsed_seconds: float = 0.0
    applied_seconds: float = 0.0
    gained_catnip: float = 0.0
    gained_scrap: float = 0.0
    gained_cats: int = 0
    gained_research: float = 0.0
    is_starved: bool = False
    starve_duration_seconds: float = 0.0
    is_capped: bool = False
    overflowed_resources: list[str] = Field(default_factory=list)
    suspicion_delta: float = 0.0
    charged_kwh: float = 0.0
    clock_anomaly: bool = False
    birth_progress: float = 0.0
    notes: list[str] = Field(default_factory=list)


class ColonyStateData(BaseModel):
    slot_id: int
    planet_id: int
    last_tick_time: int
    saved_at: int
    resources: ResourcesBlock
    power: PowerBlock
    population: PopulationBlock
    workstations: dict[str, int]
    workstation_limits: dict[str, int]
    facilities: dict[str, int]
    suspicion: SuspicionBlock
    security: SecurityBlock = SecurityBlock()
    launch_silo: LaunchSiloBlock = LaunchSiloBlock()
    offline_report: OfflineReport


class SnapshotResponse(BaseModel):
    """`POST /colony/snapshot` 响应（与定稿契约一致，不做多余包装）。"""

    code: int = 200
    message: str = "SNAPSHOT_PERSISTED"
    saved_at: int


class ColonyStateEnvelope(BaseModel):
    """`GET /colony/state` 的 `{code, data}` 封装。"""

    code: int = 200
    data: ColonyStateData


# ----------------------------------------------------------------------
# 冷启动 / 工位调度 / 设施建造（Milestone 1 阶段 3 服务端）
# ----------------------------------------------------------------------
class ScavengeResult(BaseModel):
    """手点废墟（冷启动专属动作）结果。"""

    scrap: float
    scrap_max: float
    manual_scavenge_clicks: int
    clicks_left: int
    cold_start_finished: bool = False
    hint: str | None = None


class ScavengeEnvelope(BaseModel):
    code: int = 200
    data: ScavengeResult


class DispatchPolicy(BaseModel):
    """迟滞换班策略（WBS C3，防高频震荡）。"""

    enabled: bool = True
    upper: float = Field(default=0.8, ge=0.0, le=1.0, description="高位转出线（猫薄荷占比）")
    lower: float = Field(default=0.2, ge=0.0, le=1.0, description="低位回防线（猫薄荷占比）")
    shift: int = Field(default=2, ge=1, description="一次转移猫口数")


class DispatchRequest(BaseModel):
    """`POST /colony/dispatch`：调配工位分桶并写入迟滞换班策略。"""

    model_config = ConfigDict(extra="allow")

    slot: int = Field(default=1, ge=1, le=3)
    planet_id: int | None = None
    role: str = Field(description="工种 ID（job_id）：farmer / scavenger / geek / power_runner / crew")
    delta: int = Field(description="工位增减量，正数上工、负数下岗")
    policy: DispatchPolicy | None = Field(default=None, description="可选的迟滞换班策略")


class DispatchResult(BaseModel):
    role: str
    count: int
    unassigned: int
    total_cats: int
    workstations: dict[str, int]
    workstation_limits: dict[str, int]
    power_net_kw: float
    policy: dict | None = None


class DispatchEnvelope(BaseModel):
    code: int = 200
    data: DispatchResult


class BuildRequest(BaseModel):
    """`POST /facilities/build`：建造 / 升级设施。"""

    model_config = ConfigDict(extra="allow")

    slot: int = Field(default=1, ge=1, le=3)
    planet_id: int | None = None
    facility_id: str
    count: int = Field(default=1, ge=1, le=100)


class BuildResult(BaseModel):
    facility_id: str
    level: int
    count: int
    cost_paid: dict[str, float]
    resources: dict[str, float]
    caps: dict[str, float]
    total_cats: int
    unassigned: int
    cat_capacity: int
    workstation_limits: dict[str, int]
    power_net_kw: float
    narrative: str | None = None
    unlock_hint: dict | None = None
    unlocked_planets: list[str] = Field(default_factory=list)


class BuildEnvelope(BaseModel):
    code: int = 200
    data: BuildResult
